lib/algo/DMDc.h

```diff
@@ -0,0 +1,396 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2013-2023, Lawrence Livermore National Security, LLC
+ * and other libROM project developers. See the top-level COPYRIGHT
+ * file for details.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR MIT)
+ *
+ *****************************************************************************/
+
+// Description: Computes the DMDc algorithm on the given snapshot matrix and control matrix. The
+//              implemented dynamic mode decomposition with control algorithm is derived from
+//              Proctor et. al's paper "Dynamic mode decomposition with control":
+//              https://arxiv.org/abs/1409.6358
+//              This algorithm also works in the case that the first sample does
+//              not start from t = 0.0 by incorporating a time offset.
+
+#ifndef included_DMDc_h
+#define included_DMDc_h
+
+#include <vector>
+#include <complex>
+
+namespace CAROM {
+
+class Matrix;
+class Vector;
+class ComplexEigenPair;
+
+/**
+ * Class DMDc implements the DMDc algorithm on a given snapshot matrix.
+ */
+class DMDc
```

Comment:
My answer is similar as in the previous comment: It is hard to unify DMD and DMDc into a parent class. DMDc has different matrix components and different prediction methods. We had the DMD class from which AdaptiveDMD and NonuniformDMD derives, and the ParametricDMD on top of that which interpolates matrices in DMD. Right now, I have not much idea how adpative/nonuniform version of DMDc should be implemented, but parametric version of DMDc will have to interpolate different matrices than DMD.

Reply:
There is some duplicated code, e.g. the lapack solve, that could be refactored into functions called by both classes, if the classes are kept separate. I will approve the PR without refactoring, but I just want to suggest it as a way to make the code more easily maintainable.

Reply:
Thanks Dylan. I'll keep this in mind when we continue the development of DMDc.

Reply:
I look more closely and think about the refactoring and unifying classes again. With Dylan's and Eric's suggestion, the derivation might look like this: 

`DMDBase > DMD`
`DMDBase > DMD > AdaptiveDMD`
`DMDBase > DMD > NonuniformDMD`
`DMDBase > DMDc`

Some notable difference between `DMD` and `DMDc` are: 

1. additional STL Vector class for controls
2. additional sampling for controls, and additional argument `f_in` in `takeSample`
3. optional argument `B` in `train`
4. additional sampling for controls in `computeDMDcSnapshotPair`
5. additional SVD and matrix calculations in `constructDMDc`
6. additional projection on controls, additional argument `controls` in `project`
7. more terms to sum up in `predict`
8. additional arguments `dimc` or `B_tilde` in `protected` stance of `DMDc`

I now believe the classes `DMD` and `DMDc` can be unified carefully, in a separate PR. 

We may want to add `DMDc > AdaptiveDMDc` and `DMDc > NonuniformDMDc` in the future, but I vote for not worrying about these classes until there is a need for these methods. 

On the other hand, `ParametricDMD` now works for `DMD` objects, while if it is `DMDc`, we have another matrix to interpolate. I think it can be unified easily too.